### ISSUE TITLE
Fix weapons spawning besides wrong kind of ammo

### ISSUE THREAD

**Issue by schockocraft** [TTT-2/TTT2#1004]
**Is your feature request related to a problem? Please describe.**
With the automatic weapon replacement TTT2 introduces now weapons can spawn besides the wrong kind of ammo, because Weapons with differrent ammo types are grouped in the same weapon type.

This causes Pistols and Glocks to spawn besides Deagle Ammo (when replacing a Deagle), M16s besides smg ammo (when replacing a Mac10/249), and Mac10s/249s besides 9mm ammo (when replacing a M16).

Also there is no way for map makers or Server owners to define that certain maps should not have their weapons replaced. 

Generally i would say, weapons should not be replaced, if there are (at least a few?) `ttt_random_weapon` entities on the map, as that already indicates that all other weapons are meant to stay the specific weapon chosen by the map maker or spawn script creator.

Then again there *should* be a way to spawn random primaries/ppistols/grenades, instead of just a random anything, for maps created or edited with tt2 in mind, but this way should work so that the map would still be compatible with base ttt.

**Describe the solution you'd like**
1. Replace weapons only with weapons that use the same ammo type if the weapon is close to ammo (i'd imagine checking that for any weapon might cause some lag, but then again you replace any ammo anyways already, so whatever...)
0. Allow a way to Flag a map as not needing weapon replacement. Checking if a map has `ttt_random_weapon` entities seems to be the obvious choice to me.
0. Add a Server Config/Variable to set if weapons should be replaced always, only if the map doesn't have the respective flag, only if the map is not a ttt map (could check if map name starts with `ttt_`), or not at all
0. Add a way to flag weapon entities to be replaced by random weapons of the same weapon type (aka. how it happens rn), for maps without spawn script I'd suggest just using a special entity name to identify these? That way it would stay 100% compatible with old ttt. For maps using Spawn scripts idk what to suggest for this.

**Describe alternatives you've considered**
Alternatively to solution 1 you could also change close by ammo to the ammo type of the new random weapon type

**Additional context**
-

**Comment by wgetJane** [comment]
i think it would be a lot better if weapons were grouped according to their ammo type

while grouping them by "pistol", "heavy", whatever seems to make real-world sense, it doesn't really make sense for a weapon replacement system where they'd spawn next to incorrect ammo which is just really inconvenient

or outright just group them based on the exact weapon they're replacing, like spawn types can just be "shotgun", "m16", "glock", etc and it'd be more explicit and give more control over where you want certain weapon types to be spawned (like if i wanted automatic pistols to have the "glock" spawn type and semi-auto pistols to have the "pistol" spawn type)

having weapon spawn type be a bitflag or table would also be great if you want a weapon to replace different kinds of weapon spawns

**Comment by schockocraft** [comment]
mhhh if you would go that route i'd suggest keeping the weapon groups but splitting off separate ammo consuming weapons into own groups. Because imo there's actually no reason a mac10 and a Huge shouldn't replace each other.

The weapon groupd could then be
* WEAPON_TYPE_NADE
* WEAPON_TYPE_SHOTGUN
* WEAPON_TYPE_SNIPER
* WEAPON_TYPE_PISTOL
* *WEAPON_TYPE_HEAVY_PISTOL* new for Deagle
* WEAPON_TYPE_HEAVY
* *WEAPON_TYPE_ASSAULT* new for M16



Also regarding my 
> Then again there should be a way to spawn random primaries/ppistols/grenades

i have had some more thoughts, what if you'd use the weapon entitie's names for that? they get ignored by original ttt, and ttt2 could use these to derive additional information about how to replace these weapons.
for example weapons with the name `ttt2_replaceable` would still be replaced, even on a map that already has `ttt_random_weapon`s 

One could also implement further subcategories for addons that add custom sweps to use.
for example grenades could be grouped into primarily damaging (Incendiary, etc.) and primarily tactical (Disco, Smoke) grenades. An incendiary grenade with the name `ttt2_keep_similar` would then only be replaced by grenades that damage the player, aka only incendiary grenades in vanilla ttt2, but with the right addons also frag grenades and HEs.
alternatively i'd suggest splitting WEAPON_TYPE_NADE in the first place.

Idk if that would be worth an own suggestion tho, as it is already quite away from the original point.

**Comment by wgetJane** [comment]
> imo there's actually no reason a mac10 and a Huge shouldn't replace each other

mappers typically place the m249 without any ammo nearby

**Comment by mognify** [comment]
i think there should be 3 options
• a bundle to place the weapon + its ammo
• place ammo alone
• place weapon alone

**Comment by TimGoll** [comment]
> i think there should be 3 options • a bundle to place the weapon + its ammo • place ammo alone • place weapon alone

already possible

**Comment by TimGoll** [comment]
I will come back to the general issue here, but I'm very busy at the moment

**Comment by schockocraft** [comment]
> i think there should be 3 options • a bundle to place the weapon + its ammo • place ammo alone • place weapon alone

Yea that's already possible. there's a random ammo entity that will not have a weapon spawned, and the random weapon entity also has a configurable value to specify how much ammo it will spawen (which allows 0 aka no ammo aswell)

Also in vanilla TTT just placing specific weapons or ammo can be done however you want. just with TTT2's weapon randomization you get them being replaced aswell.


Unrelated:
I thought about it furthert, and wouldn't it be actually possible to add support for a custom keyvalue in weapon entities for ttt2 only? i think vanilla ttt would just ignore it, and there wouldn't be an error or any other issues from that.

**Comment by Spanospy** [comment]
Bumping - This is a problem that still gets discussed a bit.

Idea - can the weapon spawned influence the ammo that surrounds it?

**Comment by TimGoll** [comment]
at some point I want to rework the weapons. That includes a reduction to three ammo types. Maybe this also includes an inclusion of two new spawn types as schokocraft suggested.

I think this will most likely resolve the issue. I will create a weapon rework roadmap in the coming days/weeks where you can post your suggestions

**Comment by nike4613** [comment]
This particular problem could be at least partially solved by better spawning heuristics. Even with the ammo type changes, it might be worth improving spawning heuristics anyway to try to preserve e.g. rooms with a ton of one weapon kind.

**Comment by TimGoll** [comment]
Agreed. If I (or someone else) has the time to improve the system, this could easily be improved

**Comment by wgetJane** [comment]
i don't think it needs heuristics or anything needlessly complex

i think the current "spawn types" should just be deprecated in favour of just using the ammo type and this can be a bitfield so you can mix and match them per individual weapon if you really want to (and this can be represented as a list of checkboxes or row of toggleable buttons in the equipment editor ui)

if a mapper places a mac10 in a spot in a map, 99% of the time they're placing smg ammo nearby

so like, just respect the mapper's decision? just assign weapons that use smg ammo to mac10 spawns, assign weapons that use deagle ammo to deagle spawns, etc, simple as that

doing anything other than this is arbitrary, there's no actual logical reason why mac10 & m16 & m249 spawns should get merged with each other, like is it because they all start with the letter M? i dont get it

if there are zero custom weapons installed, then weapon spawns should be effectively untouched, so each vanilla weapon should spawn exactly where the mapper intended for them to spawn

here's some pseudocode for how you can derive the spawn type of a custom weapon:
```c++
NONE = 0
RANDOM = 1
SPECIAL = 2

// bits that correspond to each vanilla weapon
GLOCK = 4
PISTOL = 8
DEAGLE = 16
SHOTGUN = 32
RIFLE = 64
MAC10 = 256
M16 = 512
M249 = 1024
SMOKE = 2048
INCEND = 4096
DISCOM = 8192

PRIMARY = SHOTGUN | RIFLE | MAC10 | M16 | M249
SECONDARY = GLOCK | PISTOL | DEAGLE
GRENADE = SMOKE | INCEND | DISCOM

// setting "RANDOM" to this would allow stuff like custom melee weapons to spawn anywhere instead of in just the "ttt_random_weapon" spawns
//RANDOM = PRIMARY | SECONDARY | GRENADE

// vanilla weapons should already have their spawn types set to exactly the one they correspond to,
// so the deagle's spawn type should be set to "DEAGLE" not "SECONDARY",
// the smoke grenade's spawn type should be set to "SMOKE" not "GRENADE", and so on


// in some function that figures out spawn types for non-vanilla weapons:

if kind = equip1 or kind = equip2 or kind = role:
    return SPECIAL

if kind = nade:
    // there's no real way to figure out what a custom grenade is supposed to do
    return GRENADE

if ammo = pistol:
    if kind = pistol:
        return GLOCK | PISTOL
    return M16

if ammo = deagle:
    return DEAGLE

if ammo = shotgun:
    return SHOTGUN

if ammo = rifle:
    return RIFLE

if ammo = smg:
    return MAC10

if ammo = lmg:
    return M249

return RANDOM
```

**Comment by Spanospy** [comment]
Was this ever worked on?
